### PR TITLE
Feat/hcl interfaces

### PR DESCRIPTION
**Pull request by rotemtam**
This PR adds support for reading from `schemaspec.Resource` trees into structs that contain slices of interface types. 

### Example 

Consider this block of HCL:
```hcl
zoo "ramat_gan" {
	lion "simba" {
		friend = "rafiki"
	}
	parrot "iago" {
		boss = "jafar"
	}
}
```

Both `lion` and `parrot` represent animals, and we would like to read them into a struct that contains an `[]Animal` slice such as:

```go
type (
  Animal interface {
	  animal()
  }
  Parrot struct {
	  Animal
	  Boss string `spec:"boss"`
  }
  Lion struct {
	  Animal
	  Friend string `spec:"friend"`
  }
  Zoo struct {
	  Animals []Animal `spec:""`
  }
)
```

### CODE REVIEW

File: schema/schemaspec/extension.go [hunk 1]
```diff
@@ -94,6 +108,30 @@ func (r *Resource) As(target interface{}) error {
 				return err
 			}
 			delete(existingAttrs, attr.K)
+		case ft.isInterfaceSlice:
```
Comment (yonidavidson) [inline schema/schemaspec/extension.go:111]: why isInterfaceSlice is a struct attribute by isResourceSlice() requires a func?

**Review by yonidavidson — APPROVED** [review]
💯

File: schema/schemaspec/extension.go [hunk 2]
```diff
@@ -53,6 +53,20 @@ func (r registry) lookup(ext interface{}) (string, bool) {
 	return "", false
 }
 
+// implementors returns a slice of the names of the extensions that implement i.
+func (r registry) implementors(i reflect.Type) ([]string, error) {
```
Comment (a8m) [inline schema/schemaspec/extension.go:57]:
```suggestion
func (r registry) implementers(i reflect.Type) ([]string, error) {
```

implementers?

File: schema/schemaspec/extension.go [hunk 3]
```diff
@@ -388,24 +426,27 @@ func specFields(ext interface{}) []fieldTag {
 		}
 		parts := strings.Split(lookup, ",")
 		fields = append(fields, fieldTag{
-			field:  f.Name,
-			tag:    lookup,
-			isName: len(parts) > 1 && parts[1] == "name",
+			field:            f.Name,
+			tag:              lookup,
+			isName:           len(parts) > 1 && parts[1] == "name",
+			isInterfaceSlice: f.Type.Kind() == reflect.Slice && f.Type.Elem().Kind() == reflect.Interface,
```
Comment (a8m) [inline schema/schemaspec/extension.go:432]: What if it's not a slice? An array or just an interface?

File: schema/schemaspec/schemahcl/hcl_test.go [hunk 4]
```diff
@@ -143,3 +144,49 @@ func ExampleMarshal() {
...
+			Animal
+			Boss string `spec:"boss"`
+		}
+		Lion struct {
+			Animal
+			Friend string `spec:"friend"`
+		}
+		Zoo struct {
+			Animals []Animal `spec:""`
+		}
```
Comment (a8m) [inline schema/schemaspec/schemahcl/hcl_test.go:173]: Can you add a test case for something like:

```go
type U struct {
	Pet Animal `spec:""`
}
```

File: schema/schemaspec/extension.go (hunk 2 above)
Comment (rotemtam) [reply schema/schemaspec/extension.go:57]: yep

File: schema/schemaspec/schemahcl/hcl_test.go (hunk 4 above)
Comment (rotemtam) [reply schema/schemaspec/schemahcl/hcl_test.go:166]: done

File: schema/schemaspec/extension.go (hunk 3 above)
Comment (rotemtam) [reply schema/schemaspec/extension.go]: added support for just an interface. will support array in a subsequent diff

**Review by a8m — APPROVED** [review]
👍

**Pull request merged**